chain/network/src/peer_store.rs

```diff
@@ -187,21 +239,149 @@ impl PeerStore {
+                                peer_id: peer_info.id.clone(),
+                                verified: TrustLevel::Signed,
+                            },
+                        );
+
+                        // Update peer_id addr
+                        self.peer_states
+                            .entry(peer_info.id.clone())
+                            .and_modify(|peer_state| peer_state.peer_info.addr = Some(peer_addr))
+                            .or_insert_with(|| KnownPeerState::new(peer_info));
+                    }
+                    TrustLevel::Direct => {
+                        // Remove peer pointing to this address if it exists.
+                        if let Some(verified_peer) = self.addr_peers.remove(&peer_addr) {
+                            self.peer_states.entry(verified_peer.peer_id).and_modify(
+                                |peer_state| {
+                                    peer_state.peer_info.addr.take();
+                                    to_touch = Some(peer_state.peer_info.id.clone());
+                                },
+                            );
+                        }
+
+                        // If this peer already exists with a signed connection ignore this update.
+                        // Warning: This is a problem for nodes that changes its address without changing peer_id.
+                        //          It is recommended to change peer_id if address is changed.
+                        if self.peer_states.get(&peer_info.id).map_or(false, |peer_state| {
+                            peer_state.peer_info.addr.map_or(false, |current_addr| {
+                                self.addr_peers.get(&current_addr).map_or(false, |verified_peer| {
+                                    verified_peer.verified == TrustLevel::Signed
+                                })
+                            })
+                        }) {
+                            continue;
+                        }
+
+                        if let Some(peer_state) = self.peer_states.get_mut(&peer_info.id) {
+                            if let Some(cur_addr) = peer_state.peer_info.addr.take() {
+                                self.addr_peers.remove(&cur_addr);
+                            }
+                        }
+
+                        // Add new address
+                        self.addr_peers.insert(
+                            peer_addr.clone(),
+                            VerifiedPeer {
+                                peer_id: peer_info.id.clone(),
+                                verified: TrustLevel::Direct,
+                            },
+                        );
+
+                        // Update peer_id addr
+                        self.peer_states
+                            .entry(peer_info.id.clone())
+                            .and_modify(|peer_state| peer_state.peer_info.addr = Some(peer_addr))
+                            .or_insert_with(|| KnownPeerState::new(peer_info));
```

Comment:
It feels to me that adding new address and updating peer_id addr can be refactored as it is common to all three branches.

Reply:
@bowenwang1996 with much more clarity today (it was too late yesterday 😅) I could refactor this code and it is much better now. 

`add_peer` just determines when `update_peer_info` should be called.
`update_peer_info` keeps the index from peers to addr and addr to peers.
It also save new information into the store.